src/lib/utils/concepts.h

```diff
@@ -0,0 +1,100 @@
+/**
+ * Useful concepts that are available throughout the library
+ * (C) 2023 Jack Lloyd
+ *     2023 René Meusel - Rohde & Schwarz Cybersecurity
+ *
+ * Botan is released under the Simplified BSD License (see license.txt)
+ */
+
+#ifndef BOTAN_CONCEPTS_H_
+#define BOTAN_CONCEPTS_H_
+
+#include <compare>
+#include <concepts>
+#include <type_traits>
+
+namespace Botan::concepts {
+
+// TODO: C++20 use std::convertible_to<> that was not available in Android NDK
+//       as of this writing. Tested with API Level up to 33.
+template <class FromT, class ToT>
+concept convertible_to =
+  std::is_convertible_v<FromT, ToT> &&
+  requires {
+    static_cast<ToT>(std::declval<FromT>());
+  };
+
+// TODO: C++20 provides concepts like std::equality_comparable or
+//       std::three_way_comparable, but at the time of this writing, some
+//       target platforms did not ship with those (Xcode 14, Android NDK r25,
+//       emscripten)
```

Comment:
Re compiler guarantees, unlike GCC, Clang and MSVC where we draw a line in the sand and say "Version X is/will be supported", for XCode and NDK we just commit to supporting the latest version. So we could fix this at any point in the 3.x releases, assuming XCode and NDK get themselves sorted out eventually.